pallets/clusters/src/lib.rs

```diff
@@ -0,0 +1,520 @@
+				let cluster = cluster
+					.as_mut()
+					.expect("Should find the cluster");
+				cluster
+					.roles
+					.iter()
+					.position(|x| x.name == role_name_vec)
+					.map(|index| cluster.roles.remove(index));
+			});
+
+			let role_hash = blake2_128(&[&cluster_id[..], &role_name_vec.as_slice()].concat());
+			<RoleToSettings<T>>::remove(&role_hash);
+
+			Self::deposit_event(Event::RoleDeleted { role_hash });
+			log::trace!("Role deleted: {:?}", role_hash);
+
+			Ok(())
+		}
+
+		/// Create a **Member**, give it a set of existing **Role** and assign it to an existing **Cluster**.
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(1).ref_time())]
+		pub fn add_member(
+			origin: OriginFor<T>,
+			cluster_id: Hash128,
+			roles: Vec<Vec<u8>>,
+			member: T::AccountId,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			// Check that the cluster exists and the caller is the owner of the cluster
+			let cluster = <Clusters<T>>::get(cluster_id).ok_or(Error::<T>::ClusterNotFound)?;
+			ensure!(who == cluster.owner, Error::<T>::NoPermission);
+
+			// Check that the cluster does not already contain the member
+			ensure!(!cluster.members.contains(&member), Error::<T>::MemberExists);
+
+			ensure!(
+				cluster.members.len() < T::MembersLimit::get() as usize,
+				Error::<T>::MembersLimitReached
+			);
+
+			// Check that the roles for the member already exists in the cluster
+			let roles_in_cluster: Vec<Vec<u8>> =
+				cluster.roles.iter().map(|role| role.name.clone()).collect();
+
+			for role in &roles {
+				ensure!(roles_in_cluster.contains(&role), Error::<T>::RoleNotFound);
+			}
+
+			// write
+			<Clusters<T>>::mutate(&cluster_id, |cluster| {
+				let cluster = cluster
+					.as_mut()
+					.expect("Should find the cluster");
+				// Add member into the cluster
+				cluster.members.push(member.clone());
+
+				// Associate the member with its roles in the cluster
+				for role in roles {
+					cluster
+						.roles
+						.iter()
+						.position(|x| x.name == role)
+						.and_then(|index| cluster.roles.get_mut(index))
+						.map(|role| role.members.push(member.clone()));
+				}
+			});
+
+			Ok(())
+		}
+
+		/// Delete a **Member** from an existing **Cluster**.
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(1).ref_time())]
+		pub fn delete_member(
+			origin: OriginFor<T>,
+			cluster_id: Hash128,
+			member: T::AccountId,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			let cluster = <Clusters<T>>::get(cluster_id).ok_or(Error::<T>::ClusterNotFound)?;
+			ensure!(who == cluster.owner, Error::<T>::NoPermission);
+			// Check that the member is in the cluster
+			ensure!(cluster.members.contains(&member), Error::<T>::MemberNotFound);
+
+			// write
+			// Delete member from Cluster
+			<Clusters<T>>::mutate(&cluster_id, |cluster| {
+				let cluster = cluster
+					.as_mut()
+					.expect("Should find the cluster");
+
+				cluster
+					.members
```

Comment:
optimize, contains iterates...

Reply:
Could be done without `iter()`, but not sure how one can reduce the time complexity 🤔